rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala

```diff
@@ -25,34 +26,30 @@ object implicits {
       }
     }
 
-  def matchListPar(init: Cost): StorageMatch[
-    BindPattern,
-    InterpreterError,
-    ListParWithRandom,
-    ListParWithRandomAndPhlos
-  ] =
-    new StorageMatch[
-      BindPattern,
-      InterpreterError,
-      ListParWithRandom,
-      ListParWithRandomAndPhlos
-    ] {
+  def matchListPar[F[_]: Sync](
+      init: Cost
+  ): StorageMatch[F, BindPattern, ListParWithRandom, ListParWithRandomAndPhlos] =
+    new StorageMatch[F, BindPattern, ListParWithRandom, ListParWithRandomAndPhlos] {
 
       private def calcUsed(init: Cost, left: Cost): Cost = init - left
 
       def get(
           pattern: BindPattern,
           data: ListParWithRandom
-      ): Either[InterpreterError, Option[ListParWithRandomAndPhlos]] =
-        SpatialMatcher
-          .foldMatch[NonDetFreeMapWithCost, Par, Par](
-            data.pars,
-            pattern.patterns,
-            pattern.remainder
-          )
-          .runFirstWithCost(init)
-          .map {
-            case (left, resultMatch) =>
+      ): F[Option[ListParWithRandomAndPhlos]] =
+        Sync[F]
+          .delay {
+            SpatialMatcher
+              .foldMatch[NonDetFreeMapWithCost, Par, Par](
+                data.pars,
+                pattern.patterns,
+                pattern.remainder
+              )
+              .runFirstWithCost(init)
+          }
+          .flatMap {
+            case Left(err) => Sync[F].raiseError(err)
```

Comment:
I'd feel much safer if we required `_error` in here and did `_error[F].raise(err)` (should compile right away, as we provide an instance of `_error` based on any MonadError[Throwable)

Reply:
The next PR should remove this line though, so I guess we can wait as long as you point me to the deletion of that in the next PR :)

Reply:
Definitely. Will do.